src/EarnerManager.sol

```diff
@@ -0,0 +1,248 @@
+    function isEarnersListIgnored() public view returns (bool isIgnored_) {
+        return IRegistrarLike(registrar).get(EARNERS_LIST_IGNORED_KEY) != bytes32(0);
+    }
+
+    /// @inheritdoc IEarnerManager
+    function isInEarnersList(address account_) public view returns (bool isInList_) {
+        return IRegistrarLike(registrar).listContains(EARNERS_LIST_NAME, account_);
+    }
+
+    /// @inheritdoc IEarnerManager
+    function getEarnerDetails(address account_) external view returns (bool status_, uint16 feeRate_, address admin_) {
+        if (isEarnersListIgnored() || isInEarnersList(account_)) return (true, 0, address(0));
+
+        EarnerDetails storage details_ = _earnerDetails[account_];
+
+        return _isValidAdmin(details_.admin) ? (true, details_.feeRate, details_.admin) : (false, 0, address(0));
+    }
+
+    /// @inheritdoc IEarnerManager
+    function getEarnerDetails(
+        address[] calldata accounts_
+    ) external view returns (bool[] memory statuses_, uint16[] memory feeRates_, address[] memory admins_) {
+        statuses_ = new bool[](accounts_.length);
+        feeRates_ = new uint16[](accounts_.length);
+        admins_ = new address[](accounts_.length);
+
+        bool isEarnersListIgnored_ = isEarnersListIgnored();
+
+        for (uint256 index_; index_ < accounts_.length; ++index_) {
+            if (isEarnersListIgnored_) {
+                statuses_[index_] = true;
+                continue;
+            }
+
+            address account_ = accounts_[index_];
+
+            if (isInEarnersList(account_)) {
+                statuses_[index_] = true;
+                continue;
+            }
+
+            EarnerDetails storage details_ = _earnerDetails[account_];
+
+            if (!_isValidAdmin(details_.admin)) continue;
+
+            statuses_[index_] = true;
+            feeRates_[index_] = details_.feeRate;
+            admins_[index_] = details_.admin;
+        }
+    }
+
+    /// @inheritdoc IEarnerManager
+    function isAdmin(address account_) public view returns (bool isAdmin_) {
+        // TODO: Consider transient storage for memoizing this check.
+        return IRegistrarLike(registrar).listContains(ADMINS_LIST_NAME, account_);
+    }
+
+    /* ============ Internal Interactive Functions ============ */
+
+    /**
+     * @dev   Sets the earner details for `account_`.
+     * @param account_ The account under which yield could generate.
+     * @param status_  Whether the account is an earner, according to the admin.
+     * @param admin_   The admin who set the details and who will collect the fee.
+     * @param feeRate_ The fee rate to be taken from the yield.
+     */
+    function _setDetails(address account_, bool status_, address admin_, uint16 feeRate_) internal {
+        if (account_ == address(0)) revert ZeroAccount();
+        if (!status_ && (feeRate_ != 0)) revert InvalidDetails();
+        if (status_ == (admin_ == address(0))) revert InvalidDetails();
```

Comment:
not possible to have address(0), msg.sender is passed in external method calls

Reply:
Correct, but functions should not make assumptions of upstream logic and checks. With a small extra line (which just costs gasa for setting details), this fucntion can be audited and tested without upstream assumptions.

Reply:
`_setDetails` changed over time of coding this up, and it appears now that we don't even need the upstream logic to pass the calling adin, and can just use `msg.sender`.